clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -472,14 +475,63 @@ static Value *EmitISOVolatileStore(CodeGenFunction &CGF, const CallExpr *E) {
   return Store;
 }
 
+static StringRef
+convertFPAccuracy(LangOptions::FPAccuracyKind FPAccuracy) {
+  StringRef AccuracyVal;
+  switch (FPAccuracy) {
+  case LangOptions::FPA_Default:
+    AccuracyVal = "fpaccuracy.default";
+    break;
+  case LangOptions::FPA_High:
+    AccuracyVal = "fpaccuracy.high";
+    break;
+  case LangOptions::FPA_Medium:
+    AccuracyVal = "fpaccuracy.medium";
+    break;
+  case LangOptions::FPA_Low:
+    AccuracyVal = "fpaccuracy.low";
+    break;
+  case LangOptions::FPA_Sycl:
+    AccuracyVal = "fpaccuracy.sycl";
+    break;
+  case LangOptions::FPA_Cuda:
+    AccuracyVal = "fpaccuracy.cuda";
+    break;
+  }
+  return AccuracyVal;
+}
+
 // Emit a simple mangled intrinsic that has 1 argument and a return type
 // matching the argument type. Depending on mode, this may be a constrained
 // floating-point intrinsic.
 static Value *emitUnaryMaybeConstrainedFPBuiltin(CodeGenFunction &CGF,
                                 const CallExpr *E, unsigned IntrinsicID,
-                                unsigned ConstrainedIntrinsicID) {
+                                unsigned ConstrainedIntrinsicID,
+    unsigned FPAccuracyIntrinsicID = Intrinsic::not_intrinsic) {
+  llvm::TargetLibraryInfoImpl TLII(CGF.getTarget().getTriple());
   llvm::Value *Src0 = CGF.EmitScalarExpr(E->getArg(0));
-
+  if (FPAccuracyIntrinsicID != Intrinsic::not_intrinsic) {
+    LangOptions::FPAccuracyKind FPAccuracy = CGF.getLangOpts().getFPAccuracy();
+    // Enter this part of the condition only when TLI.isFPACCuracyAvailable is
+    // true, implying FPAccuary != LangOptions::FPA_Default.
+    if (FPAccuracy == LangOptions::FPA_Default)
+      CGF.CGM.Error(E->getExprLoc(), "FP accuracy other than defaut is expected");
+    StringRef FPAccuracyVal = convertFPAccuracy(FPAccuracy);
+    auto *AccuracyMDS = MDString::get(CGF.Builder.getContext(), FPAccuracyVal);
+    auto *AccuracyMD =
+        MetadataAsValue::get(CGF.Builder.getContext(), AccuracyMDS);
+    Function *F = CGF.CGM.getIntrinsic(FPAccuracyIntrinsicID, Src0->getType());
+    // TODO: For now, the IR generated for cos function for example is of this form:
+    // call float @llvm.experimental.fpaccuracy.cos.f32(float %0, metadata !"fpaccuracy.value")
+    // But the final goal is to generate this IR:
+    // call float @llvm.experimental.fpaccuracy.cos.f32(float %0, float ulp)
+    // TODO:
+    // TLI.getFPAccuracy should return a string of the form "float ulp" depending
+    // on the function (IntrinsicID) and the command line accuracy (FPAccuracyIntrinsicID)
+    // FPAccuracyVal = TLI.getFPAccuracy(F->getName(), FPAccuracyIntrinsicID)
+    // auto *AccuracyMDS = MDString::get(CGF.Builder.getContext(), FPAccuracyVal);
+    return CGF.Builder.CreateCall(F, {Src0, AccuracyMD});
```

Comment:
I know the mock-up I shared used a metadata argument with a string like you have hear, but I ended up switching that to represent the accuracy as a numeric value using a callsite attribute. See PR #8134.

Reply:
Yes, I knew that was temporary. I was shooting for this IR:

  call float @llvm.experimental.fpaccuracy.cos.f32(float %0, "float ulp-value")

but we need to have a way to calculate the ulp error. From you message in Team and comments below, it looks like the FE needs to compute this ulp error without going through the TargetLibraryInfoImp? So, it would be a new function depending on the library function name and the fp-accuracy option given in the command line (or FPA_default if no option is used)?

Do you mean fpbuiltin-max-error? This attribute still needs the ulp error value. 

So, this is the IR you are expecting?
  call float @llvm.experimental.fpaccuracy.cos.f32(float %0) #10
  attributes #0 = { "fpbuiltin-max-error"="2.5" }

If yes, the FE needs to generate a new attribute to mark the function and calculate the ulp error value (the way I describe above).

Reply:
I think the mapping still needs to come from the backend because we'll want other languages (like Fortran) to be able to use the same functionality. I'll think about this and try to come up with a proposal. If you want to put a placeholder function in the front end for now, that's fine.